File: modules/vigilance-sensor/android/src/main/java/expo/modules/vigilancesensor/VigilanceSensorModule.kt
package expo.modules.vigilancesensor

import android.content.Context
import android.content.Intent
import androidx.core.content.ContextCompat
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class VigilanceSensorModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("VigilanceSensor")

    Function("isAvailable") {
      true
    }

    AsyncFunction("startMonitoring") { graceMs: Double, threshold: Double, audioUri: String ->
      val context = appContext.reactContext
        ?: throw Exceptions.ReactContextLost()

      VigilanceMonitoringService.prepareStart(
        context = context,
        graceMs = graceMs.toLong(),
        threshold = threshold.toFloat(),
        audioUri = audioUri
      )

      val intent = Intent(context, VigilanceMonitoringService::class.java).apply {
        action = VigilanceMonitoringService.ACTION_START
        putExtra(VigilanceMonitoringService.EXTRA_GRACE_MS, graceMs.toLong())
        putExtra(VigilanceMonitoringService.EXTRA_THRESHOLD, threshold.toFloat())
        putExtra(VigilanceMonitoringService.EXTRA_AUDIO_URI, audioUri)
      }
      ContextCompat.startForegroundService(context, intent)
      VigilanceMonitoringService.readStatus(context)
    }

    AsyncFunction("stopMonitoring") {
      val context = appContext.reactContext
        ?: throw Exceptions.ReactContextLost()
      VigilanceMonitoringService.markStopped(context)
      context.stopService(Intent(context, VigilanceMonitoringService::class.java))
      VigilanceMonitoringService.readStatus(context)
    }

    AsyncFunction("getStatus") {
      val context = appContext.reactContext
        ?: throw Exceptions.ReactContextLost()
      VigilanceMonitoringService.readStatus(context)
    }
  }
}

File: modules/vigilance-sensor/src/VigilanceSensorModule.ts
import { NativeModule, requireOptionalNativeModule } from 'expo';

export type NativeVigilanceStatus = {
  state: 'stopped' | 'arming' | 'armed' | 'alarming' | 'error';
  x: number;
  y: number;
  z: number;
  motion: number;
  threshold: number;
  armedAt: number;
  triggeredAt: number;
  error: string;
};

declare class VigilanceSensorModule extends NativeModule {
  isAvailable(): boolean;
  startMonitoring(
    graceMs: number,
    threshold: number,
    audioUri: string
  ): Promise<NativeVigilanceStatus>;
  stopMonitoring(): Promise<NativeVigilanceStatus>;
  getStatus(): Promise<NativeVigilanceStatus>;
}

export default requireOptionalNativeModule<VigilanceSensorModule>('VigilanceSensor');
